File: Solution/CSharp/107.BinaryTreeLevelOrderTraversalII.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<int>> LevelOrderBottom(TreeNode root) {
        List<List<int>> result = new List<List<int>>();
        if (root == null) return result;
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        while (queue.Count > 0) {
            List<int> curLevel = new List<int>();
            int count = queue.Count;
            for (int i = 0; i < count; i++) {
                TreeNode curNode = queue.Dequeue();
                curLevel.Add(curNode.val);
                if (curNode.left != null) queue.Enqueue(curNode.left);
                if (curNode.right != null) queue.Enqueue(curNode.right);
            }
        }
        result.Reverse();
        return result;
    }
}

File: Solution/Python/107. Binary Tree Level Order Traversal II.py
import sys
sys.path.append(r'F:\xjzspace\space\LeetCodeSolution\Helper\Python')
from Tree import *

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        res = []
        if root is None:
            return res

        queue = [root]
        while queue and queue[0]:
            level = []
            count = len(queue)
            for i in range(count):
                node = queue.pop(0)
                level.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            if level:
                res.append(level)
        return res[::-1]


root = Tree.CreateTree([3, 9, 20, None, None, 15, 7])
s = Solution()
res = s.levelOrderBottom(root)
print(res)
